### PR TITLE
Autograd for NDArray

### PR DESCRIPTION
**Pull request by ZihengJiang**
## Brief Introduction

Bring the idea from [MinPy](https://github.com/dmlc/minpy). Autograd for NDArray will allow users compute gradient for imperative programs easily:

``` python
from mxnet.autograd import grad
import mxnet.ndarray as nd
import random

def foo(x):
    r = random.randint(0, 1)
    if r % 2:
        return x**2
    else:
        return x/3

for i in range(10):
    inputs = nd.array([[1, 2, 3], [4, 5, 6]])
    grad_func = grad(foo)
    grad_val  = grad_func(inputs)
```

## Progress

- [x] Support for stateless ndarray operators
- [x] Support for ndarray operators with state

## Future
- Partial Gradient
- Support for hybrid of symbol and ndarray

@jermainewang @mli

### CODE REVIEW

File: include/mxnet/c_api.h [hunk 1]
```diff
@@ -516,6 +516,19 @@ MXNET_DLL int MXImperativeInvoke(AtomicSymbolCreator creator,
                                  const char **param_keys,
                                  const char **param_vals);
 
+MXNET_DLL int MXAutogradStartRecording();
+MXNET_DLL int MXAutogradStopRecording();
```
Comment (piiswrong) [inline include/mxnet/c_api.h]: merge these two into one

File: include/mxnet/c_api.h [hunk 2]
```diff
@@ -516,6 +516,19 @@ MXNET_DLL int MXImperativeInvoke(AtomicSymbolCreator creator,
...
+MXNET_DLL int MXAutogradMarkForRecord(mx_uint num_arr,
+                                      NDArrayHandle *arrays);
+MXNET_DLL int MXAutogradUnMarkForRecord(mx_uint num_arr,
```
Comment (piiswrong) [inline include/mxnet/c_api.h]: merge these two to one.

side note: why do we need to mark individual ndarrays?

File: src/ndarray/autograd.cc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+#include <mxnet/operator.h>
+#include <mxnet/executor.h>
+#include <iostream>
+#include "./autograd.h"
+
+namespace mxnet {
+namespace ndarray {
+
+AutogradRuntime* AutogradRuntime::instance_ = new AutogradRuntime();
```
Comment (piiswrong) [inline src/ndarray/autograd.cc]: should be thread local

File: src/ndarray/autograd.cc [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  std::vector<NDArray> grads;
+  std::vector<OpReqType> grad_reqs;
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    NDArray grad(inputs[i].shape(), inputs[i].ctx());
+    grad = static_cast<real_t>(1.0);
+    grads.push_back(grad);
+    grad_reqs.push_back(OpReqType::kWriteTo);
+  }
+
+  Executor *exec = Executor::Bind(sym,
```
Comment (piiswrong) [inline src/ndarray/autograd.cc]: so you went with the binding backward graph approach?

File: src/c_api/c_api_ndarray.cc [hunk 5]
```diff
@@ -278,6 +281,10 @@ int MXImperativeInvoke(AtomicSymbolCreator creator,
     }
   }
 
+  if (AutogradRuntime::Get()->IsRecording()) {
+    AutogradRuntime::Get()->RecordOp(op, attrs, ndinputs, ndoutputs);
```
Comment (piiswrong) [inline src/c_api/c_api_ndarray.cc]: you need the Operator object created here https://github.com/dmlc/mxnet/pull/5129/files#diff-d63cbc91852e426d335d0dc65b25c9c6R231 for backward

File: src/ndarray/autograd.cc [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                                  ctx,
+                                  ctx_map,
+                                  inputs,
+                                  grads,
+                                  grad_reqs,
+                                  aux_states);
+
+  exec->Forward(true);
```
Comment (piiswrong) [inline src/ndarray/autograd.cc]: Doing forward again is wasteful

File: src/ndarray/autograd.cc (hunk 4 above)
Comment (ZihengJiang) [inline src/ndarray/autograd.cc]: Nop, see [TODOs](https://github.com/dmlc/mxnet/pull/5129/files/5c24b3de4a517bb5e819e2267cce9ac47ac518d8#diff-377ece7abda57acc55d9c3a7892e61a7R70). Just raise the PR for MinPy meeting and tracking the progress. Many places will be rewriten.

Comment (piiswrong) [inline src/ndarray/autograd.cc]: Ok. Just be careful with dropout before this is fixed. This will give wrong result

**Comment by piiswrong** [comment]
Can one of the admins verify this patch?

File: include/mxnet/ndarray.h [hunk 7]
```diff
@@ -414,6 +431,8 @@ class NDArray {
   size_t offset_;
   /*! \brief type of data */
   int dtype_ = -1;
+
+  nnvm::NodeEntry entry_{nnvm::Node::Create(), 0, 0};
```
Comment (piiswrong) [inline include/mxnet/ndarray.h]: how much overhead does create cost?

File: python/mxnet/autograd.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    handles = []
+    for arr in arrays:
+        handles.append(arr.handle)
+    check_call(_LIB.MXAutogradSetMarkForRecord(
+        len(handles),
+        c_array(NDArrayHandle, handles),
+        ctypes.c_int(mark)))
+
+def compute_gradient(inputs, outputs):
+    input_handles = []
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: why do you need set_mark if you need to specify input here.

File: python/mxnet/autograd.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    check_call(_LIB.MXAutogradComputeGradient(
+        len(input_handles),
+        c_array(NDArrayHandle, input_handles),
+        len(output_handles),
+        c_array(NDArrayHandle, output_handles),
+        ctypes.byref(num_grad),
+        ctypes.byref(grad_handles)))
+    return [NDArray(NDArrayHandle(grad_handles[i])) for i in range(num_grad.value)]
+
+def grad_and_loss(func, argnum=0):
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: grad_and_loss & grad should be in minpy.

File: python/mxnet/autograd.py (hunk 8 above)
Comment (ZihengJiang) [inline python/mxnet/autograd.py]: Inputs here is all inputs used for computing forward, but you may only want gradients of part of them. Consider:
```python

f_partial(a, b, c, d):
  e0 = a * b
  e1 = b * c
  e2 = e0 * e1
  e3 = e2 * d

grad_func = grad(f_partial, [3])
grad_vals = grad_func(a, b, c, d)
```

You only want the gradient of `d` here, but you have to give `a, b, c` in order to compute the gradient.

File: include/mxnet/ndarray.h (hunk 7 above)
Comment (ZihengJiang) [inline include/mxnet/ndarray.h]: Create a Node instance then cast it as a shared pointer (https://github.com/dmlc/nnvm/blob/master/src/core/node.cc#L40), should I profile this?

File: python/mxnet/autograd.py (hunk 8 above)
Comment (ZihengJiang) [inline python/mxnet/autograd.py]: sorry, you are right, the inputs of ComputeGradient are different with forward inputs

**Comment by yinxiaochuan** [comment]
I have tried to use the autograd to calculate the gradient of mx.nd.SpatialTransformer. However, the memory consumption in GPU was growing through training. After dozens of iterations, it leads to program crash and out of memory error. Would you please check the memory consumption of autograd?

**Comment by ZihengJiang** [comment]
@yinxiaochuan thanks for your feedback, I'll look into this.

**Comment by yinxiaochuan** [comment]
I also test the code in cpu. 
I change the test_autograd.py as follows.
```
for x in xrange(1,100000):
	    test_operator_with_state()
```
memory consumption increases and run out of memory.

**Pull request merged**